**1. Read data from CSV file using pandas**

In [ ]:
from google.colab import drive
drive.mount("/drive", force_remount=True)
import csv
import pandas as pd
trueArticle = pd.read_csv('/drive/My Drive/Colab Notebooks/IST 664/2023/True.csv')

#Extract only the text field from the data"
true = trueArticle["text"]
print(true)
len(true), type(true[0])

Mounted at /drive
0        WASHINGTON (Reuters) - The head of a conservat...
1        WASHINGTON (Reuters) - Transgender people will...
2        WASHINGTON (Reuters) - The special counsel inv...
3        WASHINGTON (Reuters) - Trump campaign adviser ...
4        SEATTLE/WASHINGTON (Reuters) - President Donal...
                               ...                        
21412    BRUSSELS (Reuters) - NATO allies on Tuesday we...
21413    LONDON (Reuters) - LexisNexis, a provider of l...
21414    MINSK (Reuters) - In the shadow of disused Sov...
21415    MOSCOW (Reuters) - Vatican Secretary of State ...
21416    JAKARTA (Reuters) - Indonesia will buy 11 Sukh...
Name: text, Length: 21417, dtype: object


(21417, str)

**2.PRE-PROCESSING**


Merge the data from all articles to one single string

In [ ]:
trueText = ''
for i in range(len(true)):
    trueText += true[i]

print(trueText[0:1000])
len(trueText)

WASHINGTON (Reuters) - The head of a conservative Republican faction in the U.S. Congress, who voted this month for a huge expansion of the national debt to pay for tax cuts, called himself a “fiscal conservative” on Sunday and urged budget restraint in 2018. In keeping with a sharp pivot under way among Republicans, U.S. Representative Mark Meadows, speaking on CBS’ “Face the Nation,” drew a hard line on federal spending, which lawmakers are bracing to do battle over in January. When they return from the holidays on Wednesday, lawmakers will begin trying to pass a federal budget in a fight likely to be linked to other issues, such as immigration policy, even as the November congressional election campaigns approach in which Republicans will seek to keep control of Congress. President Donald Trump and his Republicans want a big budget increase in military spending, while Democrats also want proportional increases for non-defense “discretionary” spending on programs that support educati

51042676

There are 51042676 characters in all the articles together

Tokenization

To separate data into individual words I am using "punkt" tokenizer which separates by spaces and special characters(non-alphabetical characters)

In [19]:
import nltk
nltk.download('punkt')
trueTokens=nltk.word_tokenize(trueText)
print(len(trueTokens), trueTokens[0:1000])

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


9319417 ['WASHINGTON', '(', 'Reuters', ')', '-', 'The', 'head', 'of', 'a', 'conservative', 'Republican', 'faction', 'in', 'the', 'U.S.', 'Congress', ',', 'who', 'voted', 'this', 'month', 'for', 'a', 'huge', 'expansion', 'of', 'the', 'national', 'debt', 'to', 'pay', 'for', 'tax', 'cuts', ',', 'called', 'himself', 'a', '“', 'fiscal', 'conservative', '”', 'on', 'Sunday', 'and', 'urged', 'budget', 'restraint', 'in', '2018', '.', 'In', 'keeping', 'with', 'a', 'sharp', 'pivot', 'under', 'way', 'among', 'Republicans', ',', 'U.S.', 'Representative', 'Mark', 'Meadows', ',', 'speaking', 'on', 'CBS', '’', '“', 'Face', 'the', 'Nation', ',', '”', 'drew', 'a', 'hard', 'line', 'on', 'federal', 'spending', ',', 'which', 'lawmakers', 'are', 'bracing', 'to', 'do', 'battle', 'over', 'in', 'January', '.', 'When', 'they', 'return', 'from', 'the', 'holidays', 'on', 'Wednesday', ',', 'lawmakers', 'will', 'begin', 'trying', 'to', 'pass', 'a', 'federal', 'budget', 'in', 'a', 'fight', 'likely', 'to', 'be', 'lin

Lower-case

To maintain uniformity and consistency among the tokens format for better understanding and processing I am using islower() which makes all token to lower case leters.

In [38]:
trueWords = [w.lower() for w in trueTokens]
print(trueWords[0:1000])

['washington', '(', 'reuters', ')', '-', 'the', 'head', 'of', 'a', 'conservative', 'republican', 'faction', 'in', 'the', 'u.s.', 'congress', ',', 'who', 'voted', 'this', 'month', 'for', 'a', 'huge', 'expansion', 'of', 'the', 'national', 'debt', 'to', 'pay', 'for', 'tax', 'cuts', ',', 'called', 'himself', 'a', '“', 'fiscal', 'conservative', '”', 'on', 'sunday', 'and', 'urged', 'budget', 'restraint', 'in', '2018', '.', 'in', 'keeping', 'with', 'a', 'sharp', 'pivot', 'under', 'way', 'among', 'republicans', ',', 'u.s.', 'representative', 'mark', 'meadows', ',', 'speaking', 'on', 'cbs', '’', '“', 'face', 'the', 'nation', ',', '”', 'drew', 'a', 'hard', 'line', 'on', 'federal', 'spending', ',', 'which', 'lawmakers', 'are', 'bracing', 'to', 'do', 'battle', 'over', 'in', 'january', '.', 'when', 'they', 'return', 'from', 'the', 'holidays', 'on', 'wednesday', ',', 'lawmakers', 'will', 'begin', 'trying', 'to', 'pass', 'a', 'federal', 'budget', 'in', 'a', 'fight', 'likely', 'to', 'be', 'linked', 't

Removing non-words(non-alphabetical words) or Getting Total words
as non-alphabetical characters don't add any weightage to the content

In [42]:
resivedtrueWords = [w for w in trueWords if w.isalpha()]
print(len(resivedtrueWords))

8033628


There 8033628 words in all the articles together

Removing stop-words

Removing words which doesn't add any weightage to the content of the sentence

In [43]:
nltk.download('stopwords')
nltk_stops = nltk.corpus.stopwords.words('english')
truecontentw = [w for w in resivedtrueWords if w not in nltk_stops] #content words
print(len(truecontentw))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


4690368


There are 4690368 words after removing stop works from the articles

Lemmatization to reduce words to their root word("lemma") despite spellings and pronounciations

In [44]:
nltk.download('wordnet')
nltk.download('omw-1.4')
wnl = nltk.WordNetLemmatizer()
type(wnl)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


nltk.stem.wordnet.WordNetLemmatizer

In [34]:
truelemma =[wnl.lemmatize(t) for t in trueTokens]

#percentage reduction achieved with lemmatization
round( 100 - (len(set(truelemma))/len(set(trueTokens)) * 100), 1)

5.2

**3. DATA ANALYSIS**

List of top50 stop words and content words by their frequencies

In [52]:
from nltk import FreqDist
truestopw = [w for w in trueWords if w in nltk_stops] #stop words

truestopwfdist = FreqDist(truestopw)
truecontentwfdist = FreqDist(truecontentw)

toptruestopw = truestopwfdist.most_common(50)
toptruecontentw = truecontentwfdist.most_common(50)

print("top 50 stop words")
for pair in toptruestopw:
    print (pair)

print("top 50 content words")
for pair in toptruecontentw:
    print (pair)

top 50 stop words
('the', 482174)
('to', 245124)
('of', 204957)
('a', 197860)
('and', 181686)
('in', 180668)
('on', 108342)
('s', 99899)
('that', 88095)
('for', 79728)
('is', 55428)
('with', 54485)
('he', 54368)
('it', 49335)
('was', 47940)
('by', 47637)
('as', 47307)
('has', 46242)
('from', 39376)
('his', 38016)
('not', 37589)
('have', 36413)
('be', 34291)
('at', 33164)
('an', 32710)
('who', 27642)
('but', 26596)
('are', 26125)
('had', 25672)
('will', 25238)
('we', 22163)
('they', 22064)
('its', 21319)
('this', 21019)
('which', 20809)
('been', 19598)
('after', 19375)
('were', 18916)
('about', 17422)
('more', 17192)
('i', 17020)
('their', 16826)
('or', 15107)
('over', 13839)
('some', 12751)
('if', 12538)
('she', 11721)
('there', 11636)
('other', 11578)
('than', 11299)
top 50 content words
('said', 99024)
('trump', 54238)
('u.s.', 36971)
('would', 31524)
('reuters', 28412)
('president', 26386)
('state', 19726)
('government', 18285)
('new', 16743)
('house', 16513)
('states', 16506)
('als

From the results, it's clear that after removing stop words and non-alphabetical characters, the frequency ditribution of 50 most common words is the words that hold meaning to the sentence/corpus and also the frequency distribution of stop words(verbose) is more than frequery distribution of content words

List of top50 bigrams by frequencies

In [61]:
import re
from nltk.collocations import *
from nltk.collocations import BigramAssocMeasures
from nltk.collocations import BigramCollocationFinder

def alpha_filter(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (pattern.match(w)):
    return True
  else:
    return False

bigram_measures = nltk.collocations.BigramAssocMeasures()
finder = BigramCollocationFinder.from_words(trueWords)
scored1 = finder.score_ngrams(bigram_measures.raw_freq)
finder.apply_word_filter(alpha_filter) #filter non-aplhabetical characters
scored2 = finder.score_ngrams(bigram_measures.raw_freq)
finder.apply_word_filter(lambda w: w in nltk_stops) #filter stop words
scored3 = finder.score_ngrams(bigram_measures.raw_freq)

print("Top 50 Bigrams by frequency")
for i in range(50):
    print (scored1[i])

print("Top 50 Bigrams by frequemcy after removing punctuations")
for i in range(50):
    print (scored2[i])

print("Top 50 Bigrams by frequency after removing punctuations and stop words")
for i in range(50):
    print (scored3[i])

Top 50 Bigrams by frequency
(('’', 's'), 0.005894574735737225)
(('.', 'the'), 0.005132295292720564)
(('of', 'the'), 0.005117380196636764)
(('in', 'the'), 0.004401670190313407)
(('.', '“'), 0.0029265779179105302)
((',', 'the'), 0.002899644902680071)
((',', '”'), 0.0026274175734383384)
(('to', 'the'), 0.0023866299791070623)
(('said', '.'), 0.0023154881898728216)
((')', '-'), 0.0022909158373318845)
(('reuters', ')'), 0.002280292855229034)
(('(', 'reuters'), 0.0022801855523795103)
((',', 'a'), 0.002196060118352897)
(('in', 'a'), 0.0019584916095073328)
((',', 'and'), 0.0019338119541168724)
(('on', 'the'), 0.0017815492106426829)
(('for', 'the'), 0.0016553610596027628)
(('the', 'united'), 0.001520695983450467)
((',', 'which'), 0.0014838911060638235)
((',', 'said'), 0.0014762726037476379)
(('and', 'the'), 0.0013028711989172713)
(('united', 'states'), 0.0012958965136982282)
((',', 'who'), 0.0012742213380944323)
(('trump', '’'), 0.0012526534653401601)
(('the', 'u.s.'), 0.0012505074083496853)
(('

From result, we can say that frequency distribution of bigram with punctuations and stops words is more than frequery distribution of bigram after removing

Top 50 bigrams by PMI Measure

In [60]:
bigram_measures = nltk.collocations.BigramAssocMeasures()
finder4 = BigramCollocationFinder.from_words(trueWords)
finder4.apply_freq_filter(5)
truepmi = finder4.score_ngrams(bigram_measures.pmi)

print("Top 50 PMI Measure")
for i in range(50):
    print (truepmi[i])

Top 50 PMI Measure
(('agua', 'bonita'), 20.82988018065579)
(('clarece', 'polke'), 20.82988018065579)
(('darz', 'aab'), 20.82988018065579)
(('dori', 'esfahani'), 20.82988018065579)
(('doxycycline', 'hyclate'), 20.82988018065579)
(('ejaz', 'ashrafi'), 20.82988018065579)
(('ettore', 'rosato'), 20.82988018065579)
(('jabha', 'shamiya'), 20.82988018065579)
(('kajo', 'keji'), 20.82988018065579)
(('lista', 'marjana'), 20.82988018065579)
(('maale', 'adumim'), 20.82988018065579)
(('marjana', 'sarca'), 20.82988018065579)
(('mohseni', 'ejei'), 20.82988018065579)
(('petroleo', 'brasileiro'), 20.82988018065579)
(('aqeel', 'al-tayyar'), 20.566845774821996)
(('beji', 'caid'), 20.566845774821996)
(('moqtada', 'al-sadr'), 20.566845774821996)
(('sinan', 'ulgen'), 20.566845774821996)
(('ballard', 'spahr'), 20.566845774821992)
(('caid', 'essebsi'), 20.566845774821992)
(('gudni', 'johannesson'), 20.566845774821992)
(('hammam', 'al-alil'), 20.566845774821992)
(('irina', 'bokova'), 20.566845774821992)
(('jaba

Bigrams from PMI are different from bigrams by frequency and bigrams by PMI have very high PMI score(in the range of 20)

Top 50 Adjective words

I am getting the top50 adjective words by using pos_tag

In [62]:
from nltk import pos_tag
nltk.download('averaged_perceptron_tagger')
pos_tags = pos_tag(trueTokens)
adjtags = [ w for w in pos_tags if w[1] == 'JJ']
adjdist = FreqDist(adjtags)
topadjdist = adjdist.most_common(50)
for i in topadjdist:
  print(i)

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


(('s', 'JJ'), 21181)
(('last', 'JJ'), 11635)
(('other', 'JJ'), 11043)
(('former', 'JJ'), 9914)
(('new', 'JJ'), 9478)
(('“', 'JJ'), 8319)
(('presidential', 'JJ'), 7666)
(('political', 'JJ'), 7495)
(('military', 'JJ'), 7093)
(('first', 'JJ'), 6397)
(('Democratic', 'JJ'), 6289)
(('many', 'JJ'), 5662)
(('foreign', 'JJ'), 5581)
(('such', 'JJ'), 5492)
(('federal', 'JJ'), 5140)
(('Republican', 'JJ'), 5126)
(('nuclear', 'JJ'), 4903)
(('Russian', 'JJ'), 4900)
(('next', 'JJ'), 4872)
(('’', 'JJ'), 4679)
(('national', 'JJ'), 4451)
(('public', 'JJ'), 4397)
(('”', 'JJ'), 3844)
(('top', 'JJ'), 3745)
(('economic', 'JJ'), 3656)
(('senior', 'JJ'), 3638)
(('international', 'JJ'), 3335)
(('American', 'JJ'), 3264)
(('major', 'JJ'), 3239)
(('recent', 'JJ'), 3065)
(('several', 'JJ'), 3005)
(('own', 'JJ'), 2989)
(('legal', 'JJ'), 2797)
(('possible', 'JJ'), 2783)
(('financial', 'JJ'), 2743)
(('clear', 'JJ'), 2683)
(('Syrian', 'JJ'), 2676)
(('local', 'JJ'), 2642)
(('conservative', 'JJ'), 2640)
(('same', 'JJ'), 

**Counts**

In [65]:
trueArticleTokens = [None]*len(true)
for i in range(len(true)):
    trueArticleTokens[i]=nltk.word_tokenize(true[i]);
print(len(trueArticleTokens))
print(trueArticleTokens[0],len(trueArticleTokens[0]))

21417
['WASHINGTON', '(', 'Reuters', ')', '-', 'The', 'head', 'of', 'a', 'conservative', 'Republican', 'faction', 'in', 'the', 'U.S.', 'Congress', ',', 'who', 'voted', 'this', 'month', 'for', 'a', 'huge', 'expansion', 'of', 'the', 'national', 'debt', 'to', 'pay', 'for', 'tax', 'cuts', ',', 'called', 'himself', 'a', '“', 'fiscal', 'conservative', '”', 'on', 'Sunday', 'and', 'urged', 'budget', 'restraint', 'in', '2018', '.', 'In', 'keeping', 'with', 'a', 'sharp', 'pivot', 'under', 'way', 'among', 'Republicans', ',', 'U.S.', 'Representative', 'Mark', 'Meadows', ',', 'speaking', 'on', 'CBS', '’', '“', 'Face', 'the', 'Nation', ',', '”', 'drew', 'a', 'hard', 'line', 'on', 'federal', 'spending', ',', 'which', 'lawmakers', 'are', 'bracing', 'to', 'do', 'battle', 'over', 'in', 'January', '.', 'When', 'they', 'return', 'from', 'the', 'holidays', 'on', 'Wednesday', ',', 'lawmakers', 'will', 'begin', 'trying', 'to', 'pass', 'a', 'federal', 'budget', 'in', 'a', 'fight', 'likely', 'to', 'be', 'linke

In [75]:
import re
def alpha_filter(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (pattern.match(w)):
    return True
  else:
    return False

def punctuations(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (not(w.isnumeric()) and pattern.match(w)):
    return True
  else:
    return False

def caps_letter(w):
  if (w !="I" and w != 'i'):
      if(w.isupper()):
          return True
      else:
          return False
  else:
    return False


def exclamation(w):
    if( '!' in w):
        return True
    else:
        return False

truetotalwords = []*len(trueArticleTokens)
truecontentwords = [] * len(trueArticleTokens)
totalcaps  = [] * len(trueArticleTokens)
totalexclaim  = [] * len(trueArticleTokens)
totalpunct = [] * len(trueArticleTokens)
for i in range(len(trueArticleTokens)):
    currtruetotalwords = 0
    currtruecontentwords = 0
    currtruecapswords = 0
    currtrueexclaimwords = 0
    currtruepunctwords = 0
    for j in  range(len(trueArticleTokens[i])):
        w = trueArticleTokens[i][j]
        if (not(punctuations(w))):
            currtruetotalwords+=1
        if (not(punctuations(w)) and w not in nltk_stops):
            currtruecontentwords+=1
        if (caps_letter(w)):
            currtruecapswords +=1
        if (exclamation(w)):
            currtrueexclaimwords +=1
        if( punctuations(w)):
            currtruepunctwords +=1
    truetotalwords.append(currtruetotalwords)
    truecontentwords.append(currtruecontentwords)
    totalcaps.append(currtruecapswords)
    totalexclaim.append(currtrueexclaimwords)
    totalpunct.append(currtruepunctwords)



In [76]:
print("Total number of words")
for i in range(len(truetotalwords)):
    print (truetotalwords[i])

print("Total number of content words")
for i in range(len(truecontentwords)):
    print (truecontentwords[i])

print("Total number of capitalized words")
for i in range(len(totalcaps)):
    print (totalcaps[i])

print("Total number of exclaimation words")
for i in range(len(totalexclaim)):
    print (totalexclaim[i])

print("Total number of punctuations")
for i in range(len(totalpunct)):
    print (totalpunct[i])

Streaming output truncated to the last 5000 lines.
69
13
39
11
143
49
38
35
48
45
11
10
46
29
13
56
23
29
18
73
22
45
17
11
33
32
9
35
110
82
72
55
23
23
46
52
13
67
71
18
62
99
121
57
73
10
23
8
15
46
23
60
15
93
26
29
54
20
73
59
32
11
44
108
54
25
64
11
23
15
11
31
19
39
84
55
46
38
70
7
29
63
26
57
110
55
49
65
77
46
52
79
106
51
23
27
12
9
50
55
35
66
13
48
9
11
11
12
12
13
11
48
46
76
51
14
55
97
57
9
88
78
89
52
64
114
85
13
7
33
62
25
75
37
46
69
14
54
48
40
37
58
22
95
14
66
8
16
31
9
30
23
9
29
18
30
30
24
56
53
30
37
102
91
78
78
33
55
9
29
59
34
22
44
17
94
54
8
62
13
65
93
43
61
59
37
48
55
94
25
15
61
41
62
43
9
15
17
47
37
51
45
19
53
54
20
37
43
42
30
11
37
42
45
14
33
39
35
66
53
58
42
52
45
52
20
81
45
44
25
54
37
67
20
57
19
64
58
53
13
51
12
39
60
34
43
34
68
13
48
27
75
18
41
21
8
15
81
38
16
89
14
68
100
11
16
39
82
23
22
42
39
77
12
80
21
29
42
47
21
94
54
56
63
56
77
56
33
39
69
71
65
70
60
37
91
73
89
13
75
21
54
23
40
21
67
23
21
39
24
36
32
8
44
12
12
25
40
3

Saving the counts to CSV file

In [77]:
csv_file = '/drive/My Drive/Colab Notebooks/IST 664/2023/True.csv'
column_name1 = "TotalWords"
column_name2 = "TotalcontentWords"
column_name3 = "TotalcapitalizedWords"
column_name4 = "TotalexclaimationWords"
column_name5 = "TotalPunctuations"
trueArticle[column_name1] = truetotalwords
trueArticle[column_name2] = truecontentwords
trueArticle[column_name3] = totalcaps
trueArticle[column_name4] = totalexclaim
trueArticle[column_name5] = totalpunct

# Save the DataFrame to the CSV file
trueArticle.to_csv(csv_file, index=False)

**4. Result Interpretation**

True article has 51042676 letters while fake article has 59815411 letters, true article have 8033628  words whereas fake article have 9579621 words in it.

Now, coming to content words, true article has 4690368 words and fake article has 5168292 content words.

Content words to total words ratio in true article = 4690368/8033628 = 0.583
CContent words to total words ratio in fake arrticle = 5168292/9579621 0.539
Clearly, content words to total words ratio in true article is more than fake article


We can also see from our results that fake article have more number of exclaimations compared to true article
